src/GraphBLAS-sharp.Backend/Matrix/CSRMatrix/Kronecker.fs

```diff
@@ -0,0 +1,376 @@
+                    mapWithValueClArray queue allocationMode columnOffset m.Columns
+
+                queue.Post(Msg.CreateFreeMsg<_>(rowOffset))
+                queue.Post(Msg.CreateFreeMsg<_>(columnOffset))
+
+                match resultMatrix with
+                | None ->
+                    Some
+                        { m with
+                              Rows = newRowIndices
+                              Columns = newColumnIndices
+                              Values = copyClArray queue allocationMode m.Values }
+
+                | Some mainMatrix ->
+
+                    let newRows, newCols, newValues =
+                        mergeDisjointCOO
+                            queue
+                            mainMatrix.Rows
+                            mainMatrix.Columns
+                            mainMatrix.Values
+                            newRowIndices
+                            newColumnIndices
+                            m.Values
+
+                    Some
+                        { mainMatrix with
+                              Rows = newRows
+                              Columns = newCols
+                              Values = newValues }
+
+    let runToCOO<'a, 'b, 'c when 'a: struct and 'b: struct and 'c: struct and 'c: equality>
+        (clContext: ClContext)
+        (op: Expr<'a option -> 'b option -> 'c option>)
+        workGroupSize
+        =
+
+        let mapWithValueToCOO =
+            Map.WithValue.runToCOO clContext op workGroupSize
+
+        let insert =
+            insertMatrixWithOffset clContext workGroupSize
+
+        fun (queue: MailboxProcessor<_>) allocationMode (matrixLeft: ClMatrix.CSR<'a>) (matrixRight: ClMatrix.CSR<'b>) ->
+            let noneClCell = None |> clContext.CreateClCell
+
+            let mapWithZero =
+                mapWithValueToCOO queue allocationMode noneClCell matrixRight
+
+            queue.Post(Msg.CreateFreeMsg<_>(noneClCell))
+
+            let mutable resultMatrix = None
+
+            let leftMatrixRows = matrixLeft.RowPointers.ToHost queue
+            let leftMatrixCols = matrixLeft.Columns.ToHost queue
+            let leftMatrixVals = matrixLeft.Values.ToHost queue
+
+            for row in 0 .. (matrixLeft.RowCount - 1) do
+                let NNZInRow =
+                    leftMatrixRows.[row + 1] - leftMatrixRows.[row]
+
+                let firstElementIndex = leftMatrixRows.[row]
+
+                let rowOffset = row * matrixRight.RowCount
+
+                for offset in 0 .. (NNZInRow - 1) do
+                    let currentColumn =
+                        leftMatrixCols.[firstElementIndex + offset]
+
+                    let numberOfZeroElements =
+                        match offset with
+                        | 0 -> currentColumn
+                        | _ ->
+                            currentColumn
+                            - leftMatrixCols.[firstElementIndex + offset - 1]
+                            - 1
```

Comment:
Why can't we count the number of zero elements using Array.pairwise or something like that? Then use Array.map...